source/Modal/ReflectiveSubuniverse.lagda

```diff
@@ -0,0 +1,389 @@
+   dfunext fe λ x →
+   ○-rec-compute (Π B) (B x) (B-in-P x) (λ g → g x) f
+
+
+module _ {𝓤 𝓥 : _} (A A' : 𝓤 ̇) (B : A → 𝓥 ̇) (B' : A' → 𝓥 ̇) (f : A → A') (g : (x : A) → B x → B' (f x)) where
+
+ Σ-map-fwd : Σ B → Σ B'
+ Σ-map-fwd (x , y) = f x , g x y
+
+ module _ (f-equiv : is-equiv f) (g-equiv : (x : A) → is-equiv (g x)) where
+  private
+   f-hae : is-hae f
+   f-hae = equivs-are-haes f f-equiv
+
+   inv-f = pr₁ f-hae
+   inv-f-sec = pr₁ (pr₂ (pr₂ f-hae))
+   inv-f-ret = pr₁ (pr₂ f-hae)
+   inv-f-sec-coh = pr₂ (pr₂ (pr₂ f-hae))
+
+   inv-g[_] = λ x → inverse (g x) (g-equiv x)
+   inv-g-sec[_] = λ x → inverses-are-sections (g x) (g-equiv x)
+   inv-g-ret[_] = λ x → inverses-are-retractions (g x) (g-equiv x)
+
+
+  Σ-map-bwd : Σ B' → Σ B
+  pr₁ (Σ-map-bwd (x , y)) = inv-f x
+  pr₂ (Σ-map-bwd (x , y)) = inv-g[ inv-f x ] (transport⁻¹ B' (inv-f-sec x) y)
+
+  Σ-map-bwd-is-section-pr₁ : pr₁ ∘ Σ-map-fwd ∘ Σ-map-bwd ∼ pr₁
+  Σ-map-bwd-is-section-pr₁ (x , _) = inv-f-sec x
+
+  Σ-map-bwd-is-section-pr₂ : (u : Σ B') → transport B' (inv-f-sec (pr₁ u)) (pr₂ (Σ-map-fwd (Σ-map-bwd u))) ＝ pr₂ u
+  Σ-map-bwd-is-section-pr₂ (x , y) =
+   transport B' (inv-f-sec x) (g (inv-f x) (inv-g[ (inv-f x) ] (transport⁻¹ B' (inv-f-sec x) y))) ＝⟨ ap (transport B' (inv-f-sec x)) (inv-g-sec[ inv-f x ] (transport⁻¹ B' (inv-f-sec x) y)) ⟩
+   transport B' (inv-f-sec x) (transport⁻¹ B' (inv-f-sec x) y) ＝⟨ back-and-forth-transport (inv-f-sec x) ⟩
+   y ∎
+
+  Σ-map-bwd-is-section : Σ-map-fwd ∘ Σ-map-bwd ∼ id
+  Σ-map-bwd-is-section u = to-Σ-＝ (Σ-map-bwd-is-section-pr₁ u , Σ-map-bwd-is-section-pr₂ u)
+
+  Σ-map-bwd-is-retraction-pr₁ : pr₁ ∘ Σ-map-bwd ∘ Σ-map-fwd ∼ pr₁
+  Σ-map-bwd-is-retraction-pr₁ (x , y) = inv-f-ret x
+
+  Σ-map-bwd-is-retraction-pr₂ : (u : Σ B) → transport B (inv-f-ret (pr₁ u)) (pr₂ (Σ-map-bwd (Σ-map-fwd u))) ＝ pr₂ u
+  Σ-map-bwd-is-retraction-pr₂ (x , y) =
+   transport B (inv-f-ret x) (inv-g[ inv-f (f x) ] (transport⁻¹ B' (inv-f-sec (f x)) (g x y))) ＝⟨ nat-transport inv-g[_] (inv-f-ret x) ⁻¹ ⟩
+   inv-g[ x ] (transport (B' ∘ f) (inv-f-ret x) (transport B' (inv-f-sec (f x) ⁻¹) (g x y))) ＝⟨ ap inv-g[ x ] (transport-ap B' f (inv-f-ret x)) ⟩
+   inv-g[ x ] (transport B' (ap f (inv-f-ret x)) (transport B' (inv-f-sec (f x) ⁻¹) (g x y))) ＝⟨ ap inv-g[ x ] (transport-∙ B' (inv-f-sec (f x) ⁻¹) (ap f (inv-f-ret x)) ⁻¹) ⟩
+   inv-g[ x ] (transport B' (inv-f-sec (f x) ⁻¹ ∙ ap f (inv-f-ret x)) (g x y)) ＝⟨ ap (λ - → inv-g[ x ] (transport B' - (g x y))) aux ⟩
+   inv-g[ x ] (g x y) ＝⟨ inv-g-ret[ x ] y ⟩
+   y ∎
+   where
+    aux : inv-f-sec (f x) ⁻¹ ∙ ap f (inv-f-ret x) ＝ refl
+    aux =
+     inv-f-sec (f x) ⁻¹ ∙ ap f (inv-f-ret x) ＝⟨ ap (inv-f-sec (f x) ⁻¹ ∙_) (inv-f-sec-coh x) ⟩
+     inv-f-sec (f x) ⁻¹ ∙ inv-f-sec (f x) ＝⟨ trans-sym (inv-f-sec (f x)) ⟩
+     refl ∎
+
+
+  Σ-map-bwd-is-retraction : Σ-map-bwd ∘ Σ-map-fwd ∼ id
+  Σ-map-bwd-is-retraction u = to-Σ-＝ (Σ-map-bwd-is-retraction-pr₁ u , Σ-map-bwd-is-retraction-pr₂ u)
+
+  Σ-map-fwd-is-equiv : is-equiv Σ-map-fwd
```

Comment:
Later on I'll move this to the right place; but I was curious, did anyone try proving a lemma like this before (without univalence)? It seemed like it was a lot harder than it should have been, but I wasn't sure if I took a wrong turn somewhere. I was kind of surprised that I needed to use the `equivs-are-haes` fact.

Reply:
I think what you are looking for is here: https://www.cs.bham.ac.uk/~mhe/TypeTopology/Utilities.PairFun.html
Perhaps I should move this to `UF/`?
This is actually used a lot, particularly in the code for searchable ordinals.

Reply:
@martinescardo Oh, thanks for the tip! That's just what I needed.

Reply:
It is a very useful function, and putting it in `UF` does sound like a good idea from my perspective :)